### DISCUSSION TITLE
Evaluator Compiler generate and compile the class code everytime

### DISCUSSION THREAD

**Discussion by hongv** [evrete/evrete#21] [General]
I do the test about cache 'Evaluator' and not, the performance difference between the two is 20 times. so i think this is a basic implementation to this engine.

**Comment by andbi** [comment]
Show the code please.

**Reply by hongv** [reply]
### CacheExpressionResolverProvider
<img width="842" alt="image" src="https://user-images.githubusercontent.com/1720839/181436347-2b9053e5-77fe-4b5f-b7d7-b6e42ac9142f.png">

### CacheExpressionResolver
<img width="961" alt="image" src="https://user-images.githubusercontent.com/1720839/181436420-6dc0619a-9a26-4ae4-85c1-96dc5aa4ca3c.png">


**I add ExpressionResolverProvider SPI like above, and call a rule 1000 times**

<img width="663" alt="image" src="https://user-images.githubusercontent.com/1720839/181436687-b24fe5a4-8b24-4a61-85a9-20c1c85ac5ef.png">

**Comment by andbi** [comment]
1. Why would you compile the same `Knowledge` 1000 times? The whole idea is to build it once and spawn 1000 `StatelessSession` instances instead. The relationship between the two is very much like between `.java` and `.class` files, you don't compile java files every time you need to run a Java program, you use its precompiled byte code, right?
2. The caching you propose might work in a simple scenario like yours, but please never do that in production. The `"$ai.get() < 10"` string can have different meanings in different LHS contexts, like comparing  `ArtificialIntelligence` instances instead of `AtomicInteger` ones :-)

**Reply by hongv** [reply]
about 1,In my scenario, there are many small rules for users to arrange and combine a complete rule, so it is necessary to cache the lite rule.
about 2,i have comment in code，the key need with param type info, further more mybe have a common more to generate code.

**Reply by andbi** [reply]
Now I see the problem, you should've mentioned the goal in your original post. Well, then wrapping the default resolver in a caching instance is a way to go. When doing that, you could also rely on the `RuntimeContext<?>` variable to distinguish between different rulesets, that would potentially allow you to simplify the comments.

The engine also uses small comments in code to cache expressions / avoid duplicates, but that only works for alpha-conditions and within a single ruleset.

**Comment by andbi** [comment]
On a side note, you don't need the `context.update (obj)` line because there are no other rules that would need to "see" the update.

**Reply by hongv** [reply]
I just found your work for two days, and I'm ready to study it  :-)
